tutorial/part2.md

```diff
@@ -0,0 +1,755 @@
+
+To state the obvious, games are complicated! My first 4 or 5 attempts experimented with all sorts of ways to manage state. Every one of those games started simple and eventually fell apart when adding new features became too complex.
+
+You shouldn't have to write complex code to do complex things. With ECS, complexity arises from simplicity. Follow a few simple rules; get complex behavior.
+
+For a detailed overview of ECS in practice Thomas Biskup (ADOM) gave a great talk at the 2018 Roguelike Celebration. [There be dragons: Entity Component Systems for Roguelikes](https://www.youtube.com/watch?v=fGLJC5UY2o4&feature=youtu.be)
+
+For a formal and rather dry definition we can turn to wikipedia:
+
+> ECS follows the composition over inheritance principle that allows greater flexibility in defining entities where every object in a game's scene is an entity (e.g. enemies, bullets, vehicles, etc.). Every entity consists of one or more components which contains data or state. Therefore, the behavior of an entity can be changed at runtime by systems that add, remove or mutate components. This eliminates the ambiguity problems of deep and wide inheritance hierarchies that are difficult to understand, maintain and extend. - [wikipedia](https://en.wikipedia.org/wiki/Entity_component_system)
+
+At it's core ECS is just a way to manage your application state. State is stored in components, entities are collections of those components, and systems run logic on those entities in order to add, remove, and mutate their components.
+
+As our game grows in scope I think you will find that these 3 simple rules will help to manage the underlying complexity of it all leaving us to follow our inspiration and just make a game!
+
+Enough talk, let's install geotic, create our first entity and learn how to do things the ECS way!
+
+---
+
+Before we get into it, go ahead and install geotic and start the engine.
+
+```bash
+npm install geotic
+```
+
+Next let's create a new folder `./src/state` and add a file called `ecs.js` at `./src/state/ecs.js`. Now we can import the Engine from geotic and start it up.
+
+```javascript
+import { Engine } from "geotic";
+
+const ecs = new Engine();
+
+export default ecs;
+```
+
+Currently we store all of the state of our hero in the player object. To move around we directly mutate it's position values. Right now everything is very simple and easy to understand. Unfortunately this pattern won't scale very well if we were to add an NPC, a few monsters, maybe an item or two... mutating state directly like this very quickly becomes cumbersome, complicated, and prone to bugs that are hard to diagnose.
```

Comment:
💡 it's also makes saving/loading of the game state difficult. You'll have to figure out how to build and rebuild the data (state) for saving each game piece. ECS lends itself to easily serializing and deserializing of game state because of how the state is decoupled from systems

Reply:
Nice, I'll add this. The original tcod tutorial I am basing this on devotes an entire part to saving and loading. Will be a nice break :)

And thanks for the feedback